Revisions SDL-0245 WiFi Transport: Sharing SSID and password with Mobile Proxy (#886)
smartdevicelink/sdl_evolution: proposals/0245-sharing-wifi-ssid-and-password.md

---
>* If `secondaryTransports` is `TCP_WIFI_STA_MODE`, then the WiFi network can be established with IVI as WiFi access point and the mobile device can connect using WiFi station mode.
---
>* If `secondaryTransports` is `TCP_WIFI_AP_MODE`, then the WiFi network can be established with the mobile device as the WiFi access point and the IVI can connect using WiFi station mode.
---

The core will know the supported secondary transport based on values in the `smartdevicelink.ini` file. The possible values for the supported secondary transport value need to be updated. `WiFi_AP_MODE` and `WiFi_STA_MODE` is added below for supported secondary transport values. The SDL core will use these values to communicate supported secondary transport using protocol message `Start Service Ack`.

The SDL adapters can define desired supported secondary transports in `smartdevicelink.ini` file as shown below.

```
; Comma-separated list of transports that can be used as Secondary Transport for each Primary Transport.
; Possible values are WiFi, WiFi_AP_MODE, WiFi_STA_MODE, USB and Bluetooth.

> ; Core will not suggest Secondary Transport if the value is empty.

Author: All comments in the smartdevicelink.ini file are preceded by semi-colons. I just updated the text in those comments. Reference: https://github.com/smartdevicelink/sdl_core/blob/master/src/appMain/smartDeviceLink.ini

```
>* If `SecondaryTransportForBluetooth` has value `WiFi_STA_MODE` then `secondaryTransports` should have `TCP_WIFI_STA_MODE` in protocol message `Start Service Ack`.
---
>* If `SecondaryTransportForBluetooth` has value `WiFi_AP_MODE` then `secondaryTransports` should have `TCP_WIFI_AP_MODE` in protocol message `Start Service Ack`.

Here is an example of the Service map after defining supported secondary transport.

```
[ServicesMap]
; A matrix to specify which service is allowed on which transports. The transports are listed
; in preferred order. If a transport is not listed, then the service is not allowed
; to run on the transport.
; Only video and audio services are configurable.
; If the entry of a service is completely omitted, the service will be allowed on all transports.
; Possible values are: IAP_BLUETOOTH, IAP_USB, IAP_USB_HOST_MODE, IAP_USB_DEVICE_MODE, IAP_CARPLAY, SPP_BLUETOOTH, AOA_USB,  TCP_WIFI, TCP_WIFI_AP_MODE and TCP_WIFI_STA_MODE.
; Note: this configuration is applied even if multiple-transports feature is not enabled.
; Audio service
0x0A = TCP_WIFI_STA_MODE, IAP_CARPLAY, IAP_USB_HOST_MODE, IAP_USB_DEVICE_MODE, IAP_USB, AOA_USB
; Video service

> 0x0B = TCP_WIFI_STA_MODE, IAP_CARPLAY, IAP_USB_HOST_MODE, IAP_USB_DEVICE_MODE, IAP_USB, AOA_USB

Author: All comments in the smartdevicelink.ini file are preceded by semi-colons. I just updated the text in those comments. Reference: https://github.com/smartdevicelink/sdl_core/blob/master/src/appMain/smartDeviceLink.ini

```

### HMI API changes

1. Device Info is changed to provide the apps with WiFi capabilities to IVI. IVI will receive `DeviceInfo.wifiConfigurationInfo` in `OnAppRegistered` HMI API when the app is registered using `primary transport`. 

>* If the app can connect automatically to the WiFi access point created by the IVI then the `DeviceInfo.wifiConfigurationInfo.autoJoinWiFiSupported` will be `true`.
>* If the app supports WiFi access point mode then the `DeviceInfo.wifiConfigurationInfo.canHostWiFiNetwork` will be `true`.

2. HMI can send `OnSystemRequest` notification with `RequestType` as `WIFI_CONNECTION_REQ` to inform the proxy that it can initiate an automatic WiFi connection. This notification will be sent only when a user has provided consent to share WiFi credentials, and the connected app has supported WiFi configuration and the user attempts to launch that app. 

> 3. If WiFi configurations of the IVI and the app do not match, they cannot automatically join the WiFi network. Hence, the IVI should not send `OnSystemRequest` notification with `RequestType` as `WIFI_CONNECTION_REQ`.

Author: Done.

<enum name="RequestType">
  .
  .
  .
  <element name="WIFI_CONNECTION_REQ">
    <description> this request type is sent to mobile device when the user tries to launch app on IVI, which requires secondary transport to function</description>
  </element>
</enum>
</interface>

<interface name="BasicCommunication" version="X.X.X" date="2019-XX-XX">
	.
	.
	.
  <function name="JoinWiFiNetwork" messagetype="request">
    <description>A request from the connected device to join the specified WiFi network.</description>
    <param name="wifiState" type="WiFiStateInfo" mandatory="false">
        <description>wifi state info</description>
    </param>
    <param name="ssid" type="String" minlength="1" maxlength="32" mandatory="false">
        <description>name of the WiFi ssid</description>
    </param>
    <param name="password" type="String" minlength="1" maxlength="100" mandatory="false">
        <description>password to use to connect AP</description>
    </param>
    <param name="securityType" type="WiFiSecurityType" mandatory="false">
        <description>security type of WiFi AP</description>
    </param>   
</function>

<function name="JoinWiFiNetwork" messagetype="response">

> </function>

Author: Added

  </param>
        
  <param name="resultCode" type="Result" platform="documentation" mandatory="true">
    <description>See Result</description>
    <element name="SUCCESS"/>
    <element name="UNSUPPORTED_RESOURCE">
    <element name="DISALLOWED"/>
    <element name="REJECTED">
    <element name="TOO_MANY_PENDING_REQUESTS"/>
    <element name="APPLICATION_NOT_REGISTERED"/>
    <element name="GENERIC_ERROR"/>
    <element name="USER_DISALLOWED"/>
    <element name="DATA_NOT_AVAILABLE"/>
  </param>
        
  <param name="info" type="String" maxlength="1000" mandatory="false" platform="documentation">
    <description>Provides additional human readable info regarding the result.</description>
  </param>   
</function>
```

Possible responses for the above mentioned RPCs.

1. RPC response will be `SUCCESS` if the request is successful.
2. RPC response will be `UNSUPPORTED_RESOURCE` if HMI does not support WiFi transport for SDL. The SDL Core will also send this result code in the following scenarios:
  >* If `JoinWiFiNetwork` RPC is used, when supported secondary transport received by the mobile app in `Start Service Ack` is only `TCP_WIFI_STA_MODE`.

>   >* If `GetWiFiNetworkInfo` RPC is used, when supported secondary transport received by the mobile app in `Start Service Ack`  is only `TCP_WIFI_AP_MODE`.

Author: Done.

5. The proxy should be able to check the mobile device OS version. In case platform API support to connect WiFi is not available in a mobile device, the proxy should be able to set `DeviceInfo.wifiConfigurationInfo` values to `false` at run time.

Here is an example of the WiFi configuration APIs usage to set `DeviceInfo.wifiConfigurationInfo` in Android proxy.

>* If `MultiplexTransportConfig.setCanHostWiFiNetwork( true )` and mobile device __DOES NOT__ supports WiFi transport APIs provided by platform SDK then `DeviceInfo.wifiConfigurationInfo.canHostWiFiNetwork = false`
---
>* If `MultiplexTransportConfig.setCanAutoJoinWiFiNetwork( true )` and mobile device __DOES NOT__ supports WiFi transport APIs provided by platform SDK then `DeviceInfo.wifiConfigurationInfo.canAutoJoinWiFiNetwork = false`
---
>* If `MultiplexTransportConfig.setCanHostWiFiNetwork( false )` and mobile device __DOES__ supports WiFi transport APIs provided by platform SDK then `DeviceInfo.wifiConfigurationInfo.canHostWiFiNetwork = false`
---

> >* If `MultiplexTransportConfig.setCanAutoJoinWiFiNetwork( false )` and mobile device __DOES__ supports WiFi transport APIs provided by platform SDK then `DeviceInfo.wifiConfigurationInfo.canAutoJoinWiFiNetwork= false`

Author: Done.

5. The video streaming app is not registered using `secondary transport`. 

#### **A. `WiFi` credentials cannot be shared between App and IVI**

This case is applicable when the policy table is not updated or the user disallowed it.

#### Scenario Description
1. The proxy uses WiFi mobile APIs defined above.
2. SDL Core responds with `DISALLOWED` if the policy table is not updated.
3. The user tries to launch the app on HMI.
4. HMI shows "Waiting for Permission updates" pop-up with a timeout of 5 seconds.

##### Post Conditions 

WiFi credentials will not be shared between App and IVI.

#### **B. Policy table updated and user consent needed**

This case is applicable when the policy table is updated and the user did not provide consent to share WiFi credentials.

##### Scenario Description
1. If the proxy uses WiFi mobile APIs defined above before user consent, SDL Core responds with `DISALLOWED`.
2. The user tries to launch the app on HMI.
3. HMI determines that if WiFi connection can be automatically established using `DeviceInfo.wifiConfigurationInfo`.
4. HMI requests `GetListOfPermissions` and determines if `WiFi group` permissions needed.

> 5. HMI requests user consent to share WiFi credentials with the app and informs the user that credentials can be used the app or IVI to join the WiFi network automatically.

Reviewer: ```suggestion
5. HMI requests user consent to share WiFi credentials with the app and informs the user that credentials can be used by the app or IVI to join the WiFi network automatically.
```

6. The user provides consent to share WiFi credentials with the app or IVI. 
7. The IVI shows `Waiting for WiFi Connection. Please check your device when it is safe to do so. Apps on some devices may need your permissions to join WiFi network.` pop-up with a timeout of 5 seconds.

##### Post Conditions 

1. The proxy receives `OnPermissionChange` notification.

> 2. HMI sends `OnSystemRequest` notification with `WIFI_CONNECTION_REQ` request type. On receiving `OnSystemRequest` notification with `WIFI_CONNECTION_REQ` request type, the proxy should use mobile APIs for wifi to get or share wifi credentials. 

Reviewer: ```suggestion
2. HMI sends `OnSystemRequest` notification with `WIFI_CONNECTION_REQ` request type. On receiving `OnSystemRequest` notification with `WIFI_CONNECTION_REQ` request type, the proxy should use mobile APIs for WiFi to get or share WiFi credentials.
```

2. HMI requests `GetListOfPermissions` and determines if `WiFi group` permissions needed.
3. HMI requests user consent to share WiFi credentials and informs the user that credentials can be used to join the WiFi network automatically.
4. The user did not provide consent to share WiFi credentials between the app and IVI.
5. HMI shows `Connect WiFi to use this App` pop-up with a timeout of 5 seconds.

##### **Exception 2: HMI pop-up `Waiting for WiFi Connection` timeouts**

This case would arise if the WiFi RPCs fail for any reason.

1. The user tries to launch the app on HMI.
2. HMI shows `Connect WiFi to use this App` pop-up with a timeout of 5 seconds.

#### **C. Policy table updated and user consent provided**

This case is applicable when the policy table is updated and the user has provided consent to share WiFi credentials.

##### Scenario Description
1. The user tries to launch the app on HMI.
2. HMI sends `OnSystemRequest` notification with `WIFI_CONNECTION_REQ` request type.
3. HMI shows `Waiting for WiFi Connection. Please check your device when it is safe to do so. Some devices may need your permissions to connect Car WiFi.` pop-up with a timeout of 5 seconds.

> 4. If IVI is a WiFi access point, the proxy use `GetWiFiNetworkInfo` RPC to requests WiFi status info, the SDL Core responds with `SUCCESS` and provides WiFi credentials to the app.

Reviewer: ```suggestion
4. If IVI is a WiFi access point, the proxy uses `GetWiFiNetworkInfo` RPC to request WiFi status info, then SDL Core responds with `SUCCESS` and provides WiFi credentials to the app.
```

> 5. If the mobile device is a WiFi access point, the proxy use `JoinWiFiNetwork` RPC to share WiFi status info with the IVI, the SDL Core will forward WiFi info using HMI API `JoinWiFiNetwork` to HMI and receive result code. The SDL Core responds with `SUCCESS` on a successful WiFi connection.

Reviewer: ```suggestion
5. If the mobile device is a WiFi access point, the proxy uses `JoinWiFiNetwork` RPC to share WiFi status info with the IVI, then SDL Core will forward WiFi info using HMI API `JoinWiFiNetwork` to HMI and receive result code. SDL Core responds with `SUCCESS` on a successful WiFi connection.
```

##### Scenario Description

1. The user tries to launch the app on HMI.
2. HMI shows `Connect WiFi to use this App` pop-up with a timeout of 5 seconds.

##### Post Conditions 

1. The user manually connects WiFi to use the app.

> #### **E. Automatice WiFi connection cannot be established**

Reviewer: ```suggestion
#### **E. Automatic WiFi connection cannot be established**
```
